### PR TITLE
Added "hasPackage" assertions for Class

### LINKED ISSUE #1947: Add an assertion for checking a class package
#### Summary
I'm working on a code generation module and I need to check that generated classes have the right package.
Currently, this is the best AssertJ lets me express:

```java
Class<?> aClass;
String packageName;

assertThat(aClass.getPackageName()).isEqualTo(packageName);
```
I would like a more expressive assertion such as:

#### Example

```java
assertThat(aClass).hasPackage(packageName);
```
There could be an overload for a `Package` argument as well.

### PR DESCRIPTION
**Pull request by polarene**
This PR adds the `hasPackage()` assertion for checking if a `Class` declares a given package.
For example given: 

    package one.two;
   
    class MyClass {}

this assertion succeeds:

    assertThat(MyClass.class).hasPackage("one.two");

There's also an overload with the `java.lang.Package` argument.

#### Check List:
* Fixes #1947
* Unit tests : YES
* Javadoc with a code example (on API only) : YES

### CODE REVIEW

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 1]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
...
+   * @param info contains information about the assertion.
+   * @param actual the "actual" {@code Class}.
+   * @param aPackage the package that must be declared in the class.
+   */
+  public void assertHasPackage(AssertionInfo info, Class<?> actual, Package aPackage) {
+    assertNotNull(info, actual);
+    requireNonNull(aPackage, shouldNotBeNull("package").create());
+    Package actualPackage = actual.getPackage();
+
+    if (!java.util.Objects.equals(actualPackage, aPackage)) {
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:614]: As `aPackage` is non-null already at this point, this could be:

```suggestion
    if (!aPackage.equals(actualPackage)) {
```

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 2]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
...
+   * @param info contains information about the assertion.
+   * @param actual the "actual" {@code Class}.
+   * @param packageName the package that must be declared in the class.
+   */
+  public void assertHasPackage(AssertionInfo info, Class<?> actual, String packageName) {
+    assertNotNull(info, actual);
+    requireNonNull(packageName, shouldNotBeNull("package").create());
+    Package actualPackage = actual.getPackage();
+
+    if (actualPackage == null || !actualPackage.getName().equals(packageName)) {
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:598]: The `actualPackage == null` check made me think that for completeness we should have another `hasNoPackage()` assertion, which would cover the cases mentioned in the [`getPackage()`](https://docs.oracle.com/javase/9/docs/api/java/lang/Class.html#getPackage--) javadoc: array types, primitive types and void.

However, I wouldn't include `hasNoPackage()` in this PR but as a separate change. I will create an issue to track it.

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 3]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
...
+    }
+  }
+
+  /**
+   * Verifies that the actual {@code Class} has the given {@code Package}.
+   * @param info contains information about the assertion.
+   * @param actual the "actual" {@code Class}.
+   * @param aPackage the package that must be declared in the class.
+   */
+  public void assertHasPackage(AssertionInfo info, Class<?> actual, Package aPackage) {
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:609]:
```suggestion
  public void assertHasPackage(AssertionInfo info, Class<?> actual, Package expected) {
```

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import org.assertj.core.api.ClassAssert;
+import org.assertj.core.api.ClassAssertBaseTest;
+import org.junit.jupiter.api.DisplayName;
+
+/**
+ * Tests for <code>{@link ClassAssert#hasPackage(Package)}</code>.
+ *
+ * @author Matteo Mirk
+ */
+@DisplayName("Classes hasPackage")
```
Comment (scordio) [inline src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java:14]:
```suggestion
@DisplayName("ClassAssert hasPackage(Package)")
```

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+class ClassAssert_hasPackage_Package_Test extends ClassAssertBaseTest {
+
+  static final Package PACKAGE = Package.getPackage("org.assertj.core.api");
```
Comment (scordio) [inline src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java:17]: As `Package.getPackage()` is deprecated from java 9 onwards, I'd just use a mock here:
```suggestion
  static final Package PACKAGE = mock(Package.class);
```

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import org.assertj.core.api.ClassAssert;
+import org.assertj.core.api.ClassAssertBaseTest;
+import org.junit.jupiter.api.DisplayName;
+
+/**
+ * Tests for <code>{@link ClassAssert#hasPackage(String)}</code>.
+ *
+ * @author Matteo Mirk
+ */
+@DisplayName("Classes hasPackage")
```
Comment (scordio) [inline src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java:14]:
```suggestion
@DisplayName("ClassAssert hasPackage(String)")
```

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+class ClassAssert_hasPackage_String_Test extends ClassAssertBaseTest {
```
Comment (scordio) [inline src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java:15]:
```suggestion
class ClassAssert_hasPackage_with_String_Test extends ClassAssertBaseTest {
```

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import org.assertj.core.api.ClassAssertBaseTest;
+import org.junit.jupiter.api.DisplayName;
+
+/**
+ * Tests for <code>{@link ClassAssert#hasPackage(Package)}</code>.
+ *
+ * @author Matteo Mirk
+ */
+@DisplayName("Classes hasPackage")
+class ClassAssert_hasPackage_Package_Test extends ClassAssertBaseTest {
```
Comment (scordio) [inline src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java:15]:
```suggestion
class ClassAssert_hasPackage_with_Package_Test extends ClassAssertBaseTest {
```

File: src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.stream.Stream;
+
+import org.assertj.core.internal.ClassesBaseTest;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+@DisplayName("Classes assertHasPackage")
+class Classes_assertHasPackage_Test extends ClassesBaseTest {
```
Comment (scordio) [inline src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java:20]: Could you split this into two separate test classes, one for each assertion method? (e.g., `Classes_assertHasPackage_with_Package_Test` and `Classes_assertHasPackage_with_String_Test`)

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 10]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
     classes.assertHasPublicMethods(info, actual, methodNames);
     return myself;
   }
+
+  /**
+   * Verifies that the actual {@code Class} has the given package name.
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:652]: As we are still java 8 based and `Class#getPackageName()` comes from java 9+, I'd explicitly mention the relationship with `Class#getPackage()` behavior:

```suggestion
   * Verifies that the actual {@code Class} has the given package name (as in {@link Class#getPackage()}).
   *
```

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 11]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
...
+   * @throws AssertionError if {@code actual} is {@code null}.
+   * @throws AssertionError if the actual {@code Class} doesn't declare the given package.
+   */
+  public SELF hasPackage(String packageName) {
+    classes.assertHasPackage(info, actual, packageName);
+    return myself;
+  }
+
+  /**
+   * Verifies that the actual {@code Class} has the given package.
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:679]: As we are still java 8 based and `Class#getPackageName()` comes from java 9+, I'd explicitly mention the relationship with `Class#getPackage()` behavior:

```suggestion
   * Verifies that the actual {@code Class} has the given package (as in {@link Class#getPackage()}).
   *
```

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 12]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
...
+   *
+   * // these assertions fail:
+   * assertThat(MyClass.class).hasPackage("one");
+   * assertThat(MyClass.class).hasPackage("");
+   * assertThat(MyClass.class).hasPackage("java.lang");</code></pre>
+   *
+   * @param packageName the package name which the class must declare
+   * @return {@code this} assertions object
+   * @throws AssertionError if {@code actual} is {@code null}.
+   * @throws AssertionError if the actual {@code Class} doesn't declare the given package.
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:671]:
```suggestion
   * @throws AssertionError if the actual {@code Class} doesn't declare the given package.
   *
   * @since 3.18.0
```

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 13]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
...
+   *
+   * // these assertions fail:
+   * assertThat(MyClass.class).hasPackage(Package.getPackage("one"));
+   * assertThat(MyClass.class).hasPackage(Package.getPackage(""));
+   * assertThat(MyClass.class).hasPackage(Object.class.getPackage());</code></pre>
+   *
+   * @param aPackage the package which the class must declare
+   * @return {@code this} assertions object
+   * @throws AssertionError if {@code actual} is {@code null}.
+   * @throws AssertionError if the actual {@code Class} doesn't declare the given package.
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:699]:
```suggestion
   * @throws AssertionError if the actual {@code Class} doesn't declare the given package.
   *
   * @since 3.18.0
```

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 14]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
...
+    classes.assertHasPackage(info, actual, packageName);
+    return myself;
+  }
+
+  /**
+   * Verifies that the actual {@code Class} has the given package.
+   * <p>
+   * Example:
+   * <pre><code>
+   * package one.two;
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:683]: Join lines:
```suggestion
   * <pre><code class='java'> package one.two;
```

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java [hunk 15]
```diff
@@ -647,4 +647,59 @@ public SELF hasPublicMethods(String... methodNames) {
     classes.assertHasPublicMethods(info, actual, methodNames);
     return myself;
   }
+
+  /**
+   * Verifies that the actual {@code Class} has the given package name.
+   * <p>
+   * Example:
+   * <pre><code>
+   * package one.two;
```
Comment (scordio) [inline src/main/java/org/assertj/core/api/AbstractClassAssert.java:656]: Join lines:

```suggestion
   * <pre><code class='java'> package one.two;
```

File: src/main/java/org/assertj/core/error/ShouldHavePackage.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
+package org.assertj.core.error;
+
+import java.util.StringJoiner;
+
+/**
+ * Creates an error message indicating that a {@link Class} should have a given package.
+ * @author Matteo Mirk
```
Comment (scordio) [inline src/main/java/org/assertj/core/error/ShouldHavePackage.java:7]:
```suggestion
 * Creates an error message indicating that a {@link Class} should have a given package.
 *
 * @author Matteo Mirk
```

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 17]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
   private static void classParameterIsNotNull(Class<?> clazz) {
     requireNonNull(clazz, "The class to compare actual with should not be null");
   }
+
+  /**
+   * Verifies that the actual {@code Class} has the given {@code packagename}.
+   * @param info contains information about the assertion.
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:589]:
```suggestion
   * Verifies that the actual {@code Class} has the given {@code packageName}.
   *
   * @param info contains information about the assertion.
```

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 18]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
...
+   * @param actual the "actual" {@code Class}.
+   * @param packageName the package that must be declared in the class.
+   */
+  public void assertHasPackage(AssertionInfo info, Class<?> actual, String packageName) {
+    assertNotNull(info, actual);
+    requireNonNull(packageName, shouldNotBeNull("package").create());
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:595]:
```suggestion
    requireNonNull(packageName, shouldNotBeNull("packageName").create());
```

File: src/main/java/org/assertj/core/internal/Classes.java [hunk 19]
```diff
@@ -581,4 +583,36 @@ private static void assertNotNull(AssertionInfo info, Class<?> actual) {
...
+
+  /**
+   * Verifies that the actual {@code Class} has the given {@code Package}.
+   * @param info contains information about the assertion.
+   * @param actual the "actual" {@code Class}.
+   * @param aPackage the package that must be declared in the class.
+   */
+  public void assertHasPackage(AssertionInfo info, Class<?> actual, Package aPackage) {
+    assertNotNull(info, actual);
+    requireNonNull(aPackage, shouldNotBeNull("package").create());
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Classes.java:611]:
```suggestion
    requireNonNull(aPackage, shouldNotBeNull("aPackage").create());
```

**Review by scordio — CHANGES_REQUESTED** [review]
Thanks @polarene! The PR is already in good shape, just a few changes and it would be ready for merging.

File: src/main/java/org/assertj/core/error/ShouldHavePackage.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+   * @param actual the actual value in the failed assertion.
+   * @param packageName the expected package name
+   * @return the created {@code ErrorMessageFactory}.
+   */
+  public static ErrorMessageFactory shouldHavePackage(Class<?> actual, String packageName) {
+    final Package actualPackage = actual.getPackage();
+    if (actualPackage == null) {
+      return new ShouldHavePackage(actual, packageName);
+    }
+    return new ShouldHavePackage(actual, packageName, actualPackage.getName());
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldHavePackage.java:32]: could use ternary operator:
```suggestion
    return actualPackage == null ? new ShouldHavePackage(actual, packageName) : new ShouldHavePackage(actual, packageName, actualPackage.getName());
```

File: src/main/java/org/assertj/core/internal/Classes.java (hunk 2 above)
Comment (polarene) [reply src/main/java/org/assertj/core/internal/Classes.java:598]: Oh right, didn't think about that! I tried to follow the Javadoc closely but didn't try to cover those cases. Thanks for raising the issue, maybe I'll work on it as soon as I can.

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java (hunk 7 above)
Comment (polarene) [reply src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_String_Test.java:15]: I tried to follow the other tests convention and didn't see "with" used in the name, am I wrong?

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java (hunk 15 above)
Comment (polarene) [reply src/main/java/org/assertj/core/api/AbstractClassAssert.java:656]: Ok, but doesn't this mess up the indentation? In Intellij preview it seems so.

File: src/main/java/org/assertj/core/api/AbstractClassAssert.java (hunk 10 above)
Comment (polarene) [reply src/main/java/org/assertj/core/api/AbstractClassAssert.java:652]: Fine, I thought that references to inner logic should be avoided in API docs. But if you think it help understand the assertion let's add it. Should it be "{@link Class#getPackage().getName()}" anyway?

File: src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java (hunk 9 above)
Comment (polarene) [reply src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java:20]: Sure I'll do it. About the naming convention, as I said in the other comment, I didn't find "with" to separate the arguments in other tests around, can you confirm this? Or if this is a new convention we're introducing I'll use it for sure.

File: src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java (hunk 5 above)
Comment (polarene) [reply src/test/java/org/assertj/core/api/classes/ClassAssert_hasPackage_Package_Test.java:17]: I see, since we're just testing internal interactions, this way the code will be future-proof when it's migrated past Java 8?

File: src/main/java/org/assertj/core/internal/Classes.java (hunk 3 above)
Comment (polarene) [reply src/main/java/org/assertj/core/internal/Classes.java]: On second thought, now that I'm reviewing this modification in the IDE, I don't like it very much because it's not consistent with the other method in the class. For the String variant, it's named "packageName" which is explicative and makes sense, and the related error message is constructed like `shouldNotBeNull("packageName")`. But here changing it to "expected" loses some information and it's not coherent with the other methods in the class, where this semantic is not used. I would change it back to "aPackage", "pkg", or I'm accepting better name suggestions.

File: src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java (hunk 9 above)
Comment (polarene) [reply src/test/java/org/assertj/core/internal/classes/Classes_assertHasPackage_Test.java:20]: Sorry, my bad I missed those tests following the convention. I'll rename my classes immediately, ignore the other comments!

**Comment by polarene** [comment]
Could you review the changes? I think it should be ok now, thanks.

**Comment by scordio** [comment]
Thank you @polarene and sorry for the slow feedback, I will have a look during this week.

**Comment by polarene** [comment]
Hi @scordio no problems at all, take your time! Would you mind to tag the PR with `hacktoberfest-accepted` so it counts for this Hacktoberfest? Thank you

**Comment by scordio** [comment]
`assertj-core` has now the `hacktoberfest` topic so this PR should be considered valid once it is merged.

**Comment by polarene** [comment]
Thank you very much!

**Review by scordio — APPROVED** [review]

**Comment by scordio** [comment]
Integrated, thanks @polarene!

**Comment by polarene** [comment]
Thanks for merging @scordio! Only a small disappointment is that I would have liked to be notified about those last changes in the tests because I had a rationale behind, but it's fine anyway. Thanks for your support, looking forward to contributing again to Assertj!

**Comment by scordio** [comment]
I noticed that one test was missing and I spotted some changes I'd have done around. As I promised to merge the PR by the end of the week I did them by myself, but nothing is written in stone: what would you do differently?

**Comment by joel-costigliola** [comment]
@polarene we wanted to release 3.18.0 this last week end, even if the PR was merged, we are keen to hear what you have to say.

**Comment by polarene** [comment]
Hi guys, sorry if my words didn't come out right. I really appreciate your support and hard work, in fact, congratulations on letting this PR in time for 3.18.0, great job!

All the latest changes from Stefano are ok indeed, I just wanted to point out that the modifications in the `Classes_assertHasPackage_*` tests:

- erased a specific test case I wrote, which checked the mismatch between packages that have a common prefix (like "com.foo" and "com.foo.bar")
- eliminated a parametric test where the rationale was to check against a Package instance obtained both from `Class#getPackage()` and `Package#getPackage()` -- since the latter is deprecated in Java 9+, when AssertJ migrates from 8 it should be checked against `ClassLoader.getDefinedPackage(java.lang.String)`
- replaced the Class instances used in the test cases with Object and Collection, but I think those I chose initially made the tests more entertaining (Jedi), following the spirit of this library, like the examples with LoTR characters, and more specific to the point (MyClass) as its package and package prefix were more evident for the cases
- eliminated some // GIVEN blocks which I think made the tests more readable

That's it, thanks for reading so far.

File: src/main/java/org/assertj/core/internal/Classes.java (hunk 2 above)
Comment (polarene) [reply src/main/java/org/assertj/core/internal/Classes.java:601]: > However, I wouldn't include `hasNoPackage()` in this PR but as a separate change. I will create an issue to track it.

Hey @scordio I don't know if you created this but I would like to contribute this assertion too if you don't mind, thanks.

**Comment by scordio** [comment]
Hi @polarene,

> * erased a specific test case I wrote, which checked the mismatch between packages that have a common prefix (like "com.foo" and "com.foo.bar")

Here I assumed that classes have only one package which is uniquely identified, that's why I generalized it to `should_fail_if_packageName_does_not_match`. Do you see it differently?

> * eliminated a parametric test where the rationale was to check against a Package instance obtained both from Class#getPackage() and Package#getPackage() -- since the latter is deprecated in Java 9+, when AssertJ migrates from 8 it should be checked against ClassLoader.getDefinedPackage(java.lang.String)

Here I didn't want to test how the Package instance retrieval is done, which is JDK business, unless we have a reason for it. To the best of my knowledge, we should expect the same result between `Class#getPackage()`, `Package#getPackage(String)` and the Java 9+ `ClassLoader.getDefinedPackage(String)`.

Also, I just realized we could even write:

```java
  @ParameterizedTest
  @ValueSource(strings = "java.util")
  void should_fail_if_package_does_not_match(Package aPackage) {
    ...
  }
```

and JUnit 5 will resolve it with its implicit argument converter. The latter is actually my preference and I would update the test with it.

What do you think?

> * replaced the Class instances used in the test cases with Object and Collection, but I think those I chose initially made the tests more entertaining (Jedi), following the spirit of this library, like the examples with LoTR characters, and more specific to the point (MyClass) as its package and package prefix were more evident for the cases

I preferred to use well-known JDK types to avoid jumping back and forth in the classes and understand where the Jedi class is located. In general I like the entertaining test style, but I think it's better suited for the javadoc and for [`assertj-examples`](https://github.com/assertj/assertj-examples). If there is an option to avoid the usage of custom code in the tests, I vote for it.

> * eliminated some // GIVEN blocks which I think made the tests more readable

In general I tend to shorten tests where local variables are single values and can be inline, but I see your point. I'll put them back.

**Comment by polarene** [comment]
Hi @scordio, thank you very much for answering my points, I see your reasons and generally agree on everything.

> Here I assumed that classes have only one package which is uniquely identified, that's why I generalized it to should_fail_if_packageName_does_not_match. Do you see it differently?

Basically you're right, packages are unique for a given class, but what I wanted to show with those tests was to make clear that `hasPackage()` would fail even for a "sub-package", that is one that has an existing package as a prefix. Although Java considers a sub-package to be different from its ancestor, I wanted to make it clear both for users and contributors of AssertJ that the assertion is following the language specification, in fact, this is reflected in one of the Javadoc examples. 
Tell me if it makes sense for you to include such a test or if the Javadoc suffices.

> Here I didn't want to test how the Package instance retrieval is done (which is JDK business) unless we have a reason for it. To the best of my knowledge, we should expect the same result between Class#getPackage(), Package#getPackage(String) and the Java 9+ ClassLoader.getDefinedPackage(String).

Ok maybe I confused test responsibilities, but I really wanted to assert that the instances retrieved in different ways were checked consistently, to cover corner cases, and to be future-proof whenever something changes in the next JDK versions, so those tests could signal it. I wrote it exactly to assert that we expect the same result from the various retrieval methods.

> JUnit 5 conversions
Nice, let's discuss that in another issue, it seems a feature worth investigating. Only thing is that I wanted to make the check explicit depending on how you retrieve the package, see above. But yes, let's discuss this in a separate thread.

> I preferred to use well-known JDK types [...]

Got it, I understand your reason, thanks for clarifying. Now I know for future PRs: funny stuff only in Javadoc. The thing was, after looking around the code base I found some custom classes uses for tests here and there, so I thought it was a legit use, also I didn't write Jedi, found it in ClassesBaseTest.

Thanks for restoring the GIVEN blocks, I appreciate it!
